core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -559,6 +560,26 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  case class LeaderNode(leaderId: Int, leaderEpoch: Int, node: Node)
+
+  private def getCurrentLeader(tp: TopicPartition): LeaderNode = {
+    val partitionInfoOrError = replicaManager.getPartitionOrError(tp)
+    val (leaderId, leaderEpoch) = partitionInfoOrError match {
+      case Right(x) =>
+        (x.leaderReplicaIdOpt.getOrElse(-1), x.getLeaderEpoch)
+      case Left(x) =>
+        debug(s"Unable to retrieve local leaderId and Epoch with error $x, falling back to metadata cache")
+        metadataCache.getPartitionInfo(tp.topic, tp.partition) match {
+          case Some(pinfo) => (pinfo.leader(), pinfo.leaderEpoch())
+          case None => (-1, -1)
+        }
+    }
+    val leaderNode: Node = metadataCache.getAliveBrokerNode(leaderId, config.interBrokerListenerName).getOrElse({
```

Comment:
we shouldn't be passing through the interbroker listener name, we should be using the listener used by the original request to be consistent with the metadata request.

Is it simpler if we just consult the metadata cache? In KRaft mode, the metadata cache is the source of truth for partition leadership and is updated before the partition state gets updated.

Reply:
will update the listener name.

I think the motivation for checking replica manager first is it may be faster than metadata cache.

Reply:
my point in the previous comment is that will never be the case with KRaft.

Reply:
ok, I think replica manager is the more commonly used path which may have caching benefits but I'm not sure, does that make sense?

Reply:
Hmmm -- I'm not sure I understand "more commonly used path".
ReplicaManager will have the partition if the broker hosts it. The metadatacache is meant to be a cache of all the partitions, so I don't think it loses out on "caching benefits"

The benefit of the replica manager is that it also contains the log of the partition. If the metadata cache is sufficient (which it seems to be) we should probably just use that.

Reply:
I looked more into this and I see replica manager looks up the partition from a Pool object while metadata cache looks it up in the current image and creates a new UpdateMetadataPartitionState to return. I think we can avoid an allocation using the replica manager, also since the fetch/produce paths should have recently tried to read through replica manager I think it's more likely to give an in-memory cache hit than the metadata path. It still seems better to me to try from replica manager first, what do you all think?

Reply:
the extra object allocation is not a big issue, since the new leader and new leader lookup are not done in the common case, only in erroring cases.

populating the new leader state from the `Partition` also doesn't work for cases where the partition gets deleted from the leader, for instance in cases with reassignments, so populating from the metadata cache is both more likely to have up-to-date information (in KRaft mode, which we should assume to be the default) and it handles NotLeader in more cases.

Reply:
The metadata cache having more up to date information makes sense to me, but I don't follow the deletion case, would reading from the replica manager not return NOT_LEADER_OR_FOLLOWER there?

Reply:
We would but we would actually look it up twice in that path :)

Reply:
ok, so my understanding is in the case of partition reassignments it would be better to go directly to metadata cache, but when moving leadership within the replica set it is better to go to replica manager first. I think we should prioritize moving leadership within the replica set here since it seems more common, what do you all think?